tbc/items.go

```diff
@@ -207,34 +207,96 @@ func (gm GemColor) Intersects(o GemColor) bool {
 
 type ItemActivation func(*Simulation) Aura
 
-type Equipment []Item
+type Equipment [EquipTotem + 1]Item
 
-func NewEquipmentSet(names ...string) Equipment {
-	e := Equipment{EquipTotem: Item{}}
-	for _, v := range names {
-		item, ok := ItemsByName[v]
-		if !ok {
-			fmt.Printf("Unable to find item: '%s'\n", v)
+// Structs used for looking up items/gems/enchants
+type ItemSpec struct {
+	// Only name or ID needs to be set, not both
+	Name    string
+	ID      int32
+
+	Enchant EnchantSpec
+	Gems    []GemSpec
+}
+type GemSpec struct {
+	// Only name or ID needs to be set, not both
+	Name    string
+	ID      int32
+}
+type EnchantSpec struct {
+	// Only name or ID needs to be set, not both
+	Name    string
+	ID      int32
+}
+type EquipmentSpec [EquipTotem + 1]ItemSpec
+
+func NewEquipmentSet(equipSpec EquipmentSpec) Equipment {
+	equipment := Equipment{}
+
+	for _, itemSpec := range equipSpec {
+		item := Item{}
+		if foundItem, ok := ItemsByName[itemSpec.Name]; ok {
+			item = foundItem
+		} else if foundItem, ok := ItemsByID[itemSpec.ID]; ok {
+			item = foundItem
+		} else {
+			if itemSpec.Name != "" {
+				panic("No item with name: " + itemSpec.Name)
+			} else if itemSpec.ID != 0 {
+				panic(fmt.Sprintf("No item with id: %d", itemSpec.ID))
+			}
 			continue
 		}
-		item.Gems = make([]Gem, len(item.GemSlots))
+
+		if itemSpec.Enchant.Name != "" {
+			if enchant, ok := EnchantsByName[itemSpec.Enchant.Name]; ok {
+				item.Enchant = enchant
+			} else {
+				panic("No enchant with name: " + itemSpec.Enchant.Name)
+			}
+		} else if itemSpec.Enchant.ID != 0 {
+			if enchant, ok := EnchantsByID[itemSpec.Enchant.ID]; ok {
+				item.Enchant = enchant
+			} else {
+				panic(fmt.Sprintf("No enchant with id: %d", itemSpec.Enchant.ID))
+			}
+		}
+
+		if len(itemSpec.Gems) > 0 {
+			item.Gems = make([]Gem, len(item.GemSlots))
+
+			for gemIdx, gemSpec := range itemSpec.Gems {
+				if gem, ok := GemsByName[gemSpec.Name]; ok {
+					item.Gems[gemIdx] = gem
+				} else if gem, ok := GemsByID[gemSpec.ID]; ok {
+					item.Gems[gemIdx] = gem
+				} else {
+					if gemSpec.Name == "" {
```

Comment:
When loading up stored item set I get the panic below which puts the client into a dead state. On reload I get an exception in the JS code.

Uncaught TypeError: gem is null
    Gems http://localhost:3333/ui/components/gear.js:211
    updateEquipped http://localhost:3333/ui/components/gear.js:207
    updateEquipped http://localhost:3333/ui/components/gear.js:185
    importGear http://localhost:3333/ui/ui.js:1141
    importGear http://localhost:3333/ui/ui.js:1129
    loadPako http://localhost:3333/ui/ui.js:1211
    importGear http://localhost:3333/ui/ui.js:1128
    popgear http://localhost:3333/ui/ui.js:783

Reply:
Fixed